File: apps/patter-voice-sdk/libraries/typescript/src/providers/twilio-adapter.ts
/**
 * Twilio telephony adapter — parity with Python ``TwilioAdapter``.
 *
 * This adapter talks directly to Twilio's REST API via ``fetch`` (Basic auth).
 * It intentionally avoids any dependency on the official ``twilio`` Node SDK so
 * the Patter SDK stays small.
 *
 * See also: ``libraries/python/getpatter/providers/twilio_adapter.py``.
 */
import { getLogger } from '../logger';

const TWILIO_API_BASE = 'https://api.twilio.com/2010-04-01';

/** Escape a string for safe inclusion in TwiML attribute / text content. */
function twimlEscape(s: string): string {
  return s
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&apos;');
}

/** Constructor options for {@link TwilioAdapter}. */
export interface TwilioAdapterOptions {
  /** Optional Twilio edge region (e.g. ``ie1`` for Ireland). */
  region?: string;
}

/** Options accepted by {@link TwilioAdapter.provisionNumber}. */
export interface ProvisionNumberOptions {
  /** ISO-3166-1 alpha-2 country code, e.g. ``"US"``. */
  countryCode: string;
  /** Optional North-American area code (e.g. ``"415"``). */
  areaCode?: string;
}

/** Result returned by {@link TwilioAdapter.provisionNumber}. */
export interface ProvisionNumberResult {
  readonly phoneNumber: string;
  readonly sid: string;
}

/** Options accepted by {@link TwilioAdapter.configureNumber}. */
export interface ConfigureNumberOptions {
  /** URL Twilio should hit when the number receives a call. */
  voiceUrl: string;
  /** Optional status callback URL for call lifecycle events. */
  statusCallback?: string;
}

/** Options accepted by {@link TwilioAdapter.initiateCall}. */
export interface InitiateCallOptions {
  from: string;
  to: string;
  /**
   * TwiML or absolute URL Twilio should request when the call connects.
   * Mutually exclusive with ``streamUrl`` — provide exactly one.
   */
  url?: string;
  /**
   * Optional WebSocket stream URL. When provided (and ``url`` is not), the
   * adapter auto-builds a ``<Response><Connect><Stream>`` TwiML document
   * via :meth:`generateStreamTwiml` and sends it as the ``Twiml`` form
   * parameter. Mirrors the Python adapter's ``stream_url`` convenience path.
   */
  streamUrl?: string;
  statusCallback?: string;
  /** Value accepted by Twilio's ``MachineDetection`` parameter. */
  machineDetection?: 'Enable' | 'DetectMessageEnd' | 'false';
  /** Raw extra form parameters forwarded to the Calls endpoint. */
  extraParams?: Record<string, string>;
}

/** Result returned by {@link TwilioAdapter.initiateCall}. */
export interface InitiateCallResult {
  readonly callSid: string;
}

interface TwilioAvailableNumberPayload {
  available_phone_numbers?: Array<{ phone_number?: string }>;
}

interface TwilioIncomingNumberPayload {
  sid?: string;
  phone_number?: string;
}

interface TwilioCallPayload {
  sid?: string;
}

/** Direct REST adapter for Twilio Programmable Voice & Numbers API. */
export class TwilioAdapter {
  readonly accountSid: string;
  readonly region: string | undefined;
  private readonly baseUrl: string;
  private readonly authHeader: string;

  constructor(accountSid: string, authToken: string, opts: TwilioAdapterOptions = {}) {
    if (!accountSid) throw new Error('TwilioAdapter: accountSid is required');
    if (!authToken) throw new Error('TwilioAdapter: authToken is required');
    this.accountSid = accountSid;
    this.region = opts.region;
    // Twilio's edge/region prefix looks like ``https://api.ie1.twilio.com``.
    this.baseUrl = opts.region
      ? `https://api.${opts.region}.twilio.com/2010-04-01`
      : TWILIO_API_BASE;
    this.authHeader = `Basic ${Buffer.from(`${accountSid}:${authToken}`).toString('base64')}`;
  }

  private async request<T>(
    method: 'GET' | 'POST',
    path: string,
    body?: URLSearchParams,
  ): Promise<T> {
    const url = `${this.baseUrl}/Accounts/${encodeURIComponent(this.accountSid)}${path}`;
    const headers: Record<string, string> = { Authorization: this.authHeader };
    if (body) headers['Content-Type'] = 'application/x-www-form-urlencoded';

    const response = await fetch(url, {
      method,
      headers,
      body: body ? body.toString() : undefined,
      signal: AbortSignal.timeout(30_000),
    });

    const text = await response.text();
    if (!response.ok) {
      throw new Error(`Twilio ${method} ${path} failed: ${response.status} ${text}`);
    }
    if (!text) return {} as T;
    try {
      return JSON.parse(text) as T;
    } catch (e) {
      throw new Error(`Twilio returned non-JSON response: ${String(e)}`);
    }
  }

  /**
   * Provision a local phone number in the given country.
   *
   * Lists available local numbers, then purchases the first match.
   */
  async provisionNumber(opts: ProvisionNumberOptions): Promise<ProvisionNumberResult> {
    const country = encodeURIComponent(opts.countryCode);
    const queryParts: string[] = ['PageSize=1'];
    if (opts.areaCode) queryParts.push(`AreaCode=${encodeURIComponent(opts.areaCode)}`);
    const path = `/AvailablePhoneNumbers/${country}/Local.json?${queryParts.join('&')}`;

    const available = await this.request<TwilioAvailableNumberPayload>('GET', path);
    const first = available.available_phone_numbers?.[0]?.phone_number;
    if (!first) {
      throw new Error(`TwilioAdapter: no numbers available for country ${opts.countryCode}`);
    }

    const body = new URLSearchParams({ PhoneNumber: first });
    const purchased = await this.request<TwilioIncomingNumberPayload>(
      'POST',
      '/IncomingPhoneNumbers.json',
      body,
    );
    if (!purchased.sid || !purchased.phone_number) {
      throw new Error('TwilioAdapter: malformed response from IncomingPhoneNumbers.create');
    }
    return { phoneNumber: purchased.phone_number, sid: purchased.sid };
  }

  /** Update an already-purchased number to point at our voice webhook. */
  async configureNumber(phoneNumberSid: string, opts: ConfigureNumberOptions): Promise<void> {
    if (!phoneNumberSid) throw new Error('TwilioAdapter: phoneNumberSid is required');
    const body = new URLSearchParams({
      VoiceUrl: opts.voiceUrl,
      VoiceMethod: 'POST',
    });
    if (opts.statusCallback) body.set('StatusCallback', opts.statusCallback);
    await this.request<TwilioIncomingNumberPayload>(
      'POST',
      `/IncomingPhoneNumbers/${encodeURIComponent(phoneNumberSid)}.json`,
      body,
    );
  }

  /** Place an outbound call. Returns the Twilio call SID. */
  async initiateCall(opts: InitiateCallOptions): Promise<InitiateCallResult> {
    if (!opts.url && !opts.streamUrl) {
      throw new Error('TwilioAdapter: initiateCall requires either url or streamUrl');
    }
    const body = new URLSearchParams({
      From: opts.from,
      To: opts.to,
    });
    if (opts.url) {
      body.set('Url', opts.url);
    } else if (opts.streamUrl) {
      // Mirror the Python adapter: auto-build a ``<Connect><Stream>``
      // TwiML doc and inline it via the ``Twiml`` form parameter.
      body.set('Twiml', TwilioAdapter.generateStreamTwiml(opts.streamUrl));
    }
    if (opts.statusCallback) body.set('StatusCallback', opts.statusCallback);
    if (opts.machineDetection) body.set('MachineDetection', opts.machineDetection);
    if (opts.extraParams) {
      for (const [key, value] of Object.entries(opts.extraParams)) {
        body.set(key, value);
      }
    }

    const call = await this.request<TwilioCallPayload>('POST', '/Calls.json', body);
    if (!call.sid) {
      throw new Error('TwilioAdapter: Calls.create returned no SID');
    }
    return { callSid: call.sid };
  }

  /**
   * Build a ``<Response><Connect><Stream url="...">`` TwiML document.
   *
   * ``parameters`` is forwarded as ``<Parameter name="..." value="..."/>``
   * children of ``<Stream>``. Twilio Media Streams strips query-string params
   * from the ``<Stream url=...>`` before the WS handshake, so
   * ``<Parameter>`` tags are the supported way to pre-populate
   * ``start.customParameters`` on the WS ``start`` frame. Used by the
   * inbound path to carry caller / callee through to the bridge.
   *
   * Mirrors the Python adapter's ``generate_stream_twiml``.
   */
  static generateStreamTwiml(
    streamUrl: string,
    parameters?: Record<string, string>,
  ): string {
    const esc = (s: string): string =>
      s
        .replace(/&/g, '&amp;')
        .replace(/</g, '&lt;')
        .replace(/>/g, '&gt;')
        .replace(/"/g, '&quot;')
        .replace(/'/g, '&apos;');
    const escapedUrl = esc(streamUrl);
    let paramTags = '';
    if (parameters) {
      for (const [name, value] of Object.entries(parameters)) {
        if (value == null) continue;
        paramTags += `<Parameter name="${esc(name)}" value="${esc(String(value))}"/>`;
      }
    }
    return `<?xml version="1.0" encoding="UTF-8"?><Response><Connect><Stream url="${escapedUrl}">${paramTags}</Stream></Connect></Response>`;
  }

  /**
   * TwiML that parks the CALLER leg in the warm-transfer conference.
   *
   * `startConferenceOnEnter="false"` keeps the caller on Twilio's default
   * hold music until a participant with `startConferenceOnEnter="true"` (the
   * human agent) joins; `endConferenceOnExit="true"` tears the conference
   * down if the caller hangs up while waiting. When `statusCallbackUrl` is
   * provided, conference lifecycle events (start / end / join / leave) are
   * posted there for observability.
   *
   * Mirrors the Python adapter's `generate_warm_transfer_caller_twiml`.
   */
  static generateWarmTransferCallerTwiml(
    conferenceName: string,
    statusCallbackUrl = '',
  ): string {
    const esc = twimlEscape;
    const callbackAttrs = statusCallbackUrl
      ? ` statusCallback="${esc(statusCallbackUrl)}" statusCallbackEvent="start end join leave"`
      : '';
    return (
      '<?xml version="1.0" encoding="UTF-8"?><Response><Dial>' +
      `<Conference startConferenceOnEnter="false" endConferenceOnExit="true"${callbackAttrs}>` +
      `${esc(conferenceName)}</Conference></Dial></Response>`
    );
  }

  /**
   * TwiML executed on the TARGET (human agent) leg of a warm transfer.
   *
   * Speaks the agent-provided handoff `summary` first (skipped when empty),
   * then joins the conference with `startConferenceOnEnter="true"` — which
   * starts the conference, stops the caller's hold music, and bridges the
   * two. `endConferenceOnExit="true"` ends the conference (and therefore the
   * caller leg) when the human agent hangs up.
   *
   * Mirrors the Python adapter's `generate_warm_transfer_target_twiml`.
   */
  static generateWarmTransferTargetTwiml(
    conferenceName: string,
    summary = '',
  ): string {
    const esc = twimlEscape;
    const say = summary ? `<Say>${esc(summary)}</Say>` : '';
    return (
      `<?xml version="1.0" encoding="UTF-8"?><Response>${say}<Dial>` +
      '<Conference startConferenceOnEnter="true" endConferenceOnExit="true">' +
      `${esc(conferenceName)}</Conference></Dial></Response>`
    );
  }

  /** Force-complete an in-progress call. */
  async endCall(callSid: string): Promise<void> {
    if (!callSid) throw new Error('TwilioAdapter: callSid is required');
    const body = new URLSearchParams({ Status: 'completed' });
    try {
      await this.request<TwilioCallPayload>(
        'POST',
        `/Calls/${encodeURIComponent(callSid)}.json`,
        body,
      );
    } catch (err) {
      getLogger().warn(`[TwilioAdapter] endCall failed for ${callSid}: ${String(err)}`);
      throw err;
    }
  }
}

File: apps/patter-voice-sdk/libraries/python/getpatter/providers/twilio_adapter.py
"""Twilio :class:`TelephonyProvider` — number provisioning and call control.

Async wrapper over the synchronous Twilio REST client. Sync calls are
dispatched to a thread executor so the event loop is never blocked.
"""

import asyncio
import logging
import re
from functools import partial
from twilio.rest import Client as TwilioClient
from twilio.twiml.voice_response import VoiceResponse, Connect
from getpatter.providers.base import TelephonyProvider

logger = logging.getLogger("getpatter.providers.twilio_adapter")


_PASCAL_TO_SNAKE_RE = re.compile(r"(?<!^)(?=[A-Z])")


def _to_snake_case(name: str) -> str:
    """Translate a PascalCase / camelCase Twilio param to snake_case.

    The ``twilio-python`` SDK's ``client.calls.create(**kwargs)`` accepts
    snake_case keyword arguments only — it translates them to the
    PascalCase form Twilio's REST wire protocol requires. Passing a
    PascalCase key directly raises ``TypeError: unexpected keyword
    argument``. This helper normalises both shapes so the adapter is
    robust regardless of how the caller spelled the param.
    """
    return _PASCAL_TO_SNAKE_RE.sub("_", name).lower()


class TwilioAdapter(TelephonyProvider):
    """:class:`TelephonyProvider` implementation backed by the Twilio REST API."""

    def __init__(self, account_sid: str, auth_token: str):
        self.account_sid = account_sid
        self.auth_token = auth_token
        self._twilio_client = TwilioClient(account_sid, auth_token)

    def __repr__(self) -> str:
        masked = f"{self.account_sid[:6]}..." if len(self.account_sid) > 6 else "***"
        return f"TwilioAdapter(account_sid={masked!r})"

    async def _run_sync(self, func, *args, **kwargs):
        loop = asyncio.get_running_loop()
        return await loop.run_in_executor(None, partial(func, *args, **kwargs))

    async def provision_number(self, country: str) -> str:
        """Find and purchase a local Twilio number in the given ISO country."""
        available = await self._run_sync(
            self._twilio_client.available_phone_numbers(country).local.list, limit=1
        )
        if not available:
            raise ValueError(f"No numbers for {country}")
        purchased = await self._run_sync(
            self._twilio_client.incoming_phone_numbers.create,
            phone_number=available[0].phone_number,
        )
        return purchased.phone_number

    async def configure_number(self, number: str, webhook_url: str) -> None:
        """Point Twilio's voice webhook for *number* at *webhook_url* (POST)."""
        numbers = await self._run_sync(
            self._twilio_client.incoming_phone_numbers.list, phone_number=number
        )
        if not numbers:
            raise ValueError(f"Number {number} not found")
        await self._run_sync(
            numbers[0].update, voice_url=webhook_url, voice_method="POST"
        )

    async def initiate_call(
        self,
        from_number: str,
        to_number: str,
        stream_url: str,
        extra_params: dict | None = None,
    ) -> str:
        """Place an outbound Twilio call that streams media to *stream_url*."""
        twiml = VoiceResponse()
        connect = Connect()
        connect.stream(url=stream_url)
        twiml.append(connect)
        call_kwargs: dict = {"to": to_number, "from_": from_number, "twiml": str(twiml)}
        if extra_params:
            # Defensive normalisation: the ``twilio-python`` SDK rejects
            # PascalCase kwargs (``StatusCallback``, ``MachineDetection``,
            # …) with ``TypeError: unexpected keyword argument``.
            # ``getpatter.client`` already builds the dict in snake_case
            # form; this guard catches any third-party caller (or future
            # regression) that still passes the wire-protocol spelling.
            for key, value in extra_params.items():
                call_kwargs[_to_snake_case(key)] = value
        call = await self._run_sync(self._twilio_client.calls.create, **call_kwargs)
        return call.sid

    async def end_call(self, call_id: str) -> None:
        """Hang up the named Twilio call by setting status=completed."""
        await self._run_sync(
            self._twilio_client.calls(call_id).update, status="completed"
        )

    def record_call_end_cost(self, *, duration_seconds: float, direction: str) -> None:
        """Emit ``patter.cost.telephony_minutes`` on the active span.

        Called by the embedded server's bridge cleanup once the call's
        wall-clock duration is known.
        """
        try:
            from getpatter.observability.attributes import record_patter_attrs

            record_patter_attrs(
                {
                    "patter.cost.telephony_minutes": duration_seconds / 60.0,
                    "patter.telephony": "twilio",
                    "patter.direction": direction,
                }
            )
        except Exception:  # pragma: no cover — defense in depth
            logger.debug("record_call_end_cost failed", exc_info=True)

    @staticmethod
    def generate_warm_transfer_caller_twiml(
        conference_name: str,
        status_callback_url: str = "",
    ) -> str:
        """TwiML that parks the CALLER leg in the warm-transfer conference.

        ``startConferenceOnEnter=false`` keeps the caller on Twilio's default
        hold music until a participant with ``startConferenceOnEnter=true``
        (the human agent) joins; ``endConferenceOnExit=true`` tears the
        conference down if the caller hangs up while waiting. When
        ``status_callback_url`` is provided, conference lifecycle events
        (start / end / join / leave) are posted there for observability.
        """
        from twilio.twiml.voice_response import Dial

        response = VoiceResponse()
        dial = Dial()
        conference_kwargs: dict = {
            "start_conference_on_enter": False,
            "end_conference_on_exit": True,
        }
        if status_callback_url:
            conference_kwargs["status_callback"] = status_callback_url
            conference_kwargs["status_callback_event"] = "start end join leave"
        dial.conference(conference_name, **conference_kwargs)
        response.append(dial)
        return str(response)

    @staticmethod
    def generate_warm_transfer_target_twiml(
        conference_name: str,
        summary: str = "",
    ) -> str:
        """TwiML executed on the TARGET (human agent) leg of a warm transfer.

        Speaks the agent-provided handoff ``summary`` first (skipped when
        empty), then joins the conference with ``startConferenceOnEnter=true``
        — which starts the conference, stops the caller's hold music, and
        bridges the two. ``endConferenceOnExit=true`` ends the conference
        (and therefore the caller leg) when the human agent hangs up.
        """
        from twilio.twiml.voice_response import Dial

        response = VoiceResponse()
        if summary:
            response.say(summary)
        dial = Dial()
        dial.conference(
            conference_name,
            start_conference_on_enter=True,
            end_conference_on_exit=True,
        )
        response.append(dial)
        return str(response)

    @staticmethod
    def generate_stream_twiml(
        stream_url: str,
        parameters: dict[str, str] | None = None,
    ) -> str:
        """Return TwiML that connects the inbound call to the media stream URL.

        ``parameters`` is forwarded as ``<Parameter name="..." value="..."/>``
        children of ``<Stream>``. Twilio Media Streams ignores query-string
        params on the ``<Stream url=...>`` (they are stripped before the WS
        handshake), so ``<Parameter>`` tags are the supported way to
        pre-populate ``start.customParameters`` on the WS payload. Used by
        the inbound path to carry caller / callee through to the bridge.
        """
        response = VoiceResponse()
        connect = Connect()
        stream = connect.stream(url=stream_url)
        if parameters:
            for name, value in parameters.items():
                if value is None:
                    continue
                stream.parameter(name=name, value=str(value))
        response.append(connect)
        return str(response)
